driver/sql/projection_notification_queue.go

```diff
@@ -0,0 +1,96 @@
+package sql
+
+import (
+	"context"
+	"errors"
+	"time"
+)
+
+// Ensure the NotificationQueue.Queue is a ProjectionTrigger
+var _ ProjectionTrigger = (&NotificationQueue{}).Queue
+
+// Ensure the NotificationQueue.ReQueue is a ProjectionTrigger
+var _ ProjectionTrigger = (&NotificationQueue{}).ReQueue
+
+type (
+	// NotificationQueueInterface describes a smart queue for projection notifications
+	NotificationQueueInterface interface {
+		Channel() chan *ProjectionNotification
+		Close()
+		Empty() bool
+		Open(chan struct{})
+		PutBack(*ProjectionNotification)
+		Queue(context.Context, *ProjectionNotification) error
+		ReQueue(context.Context, *ProjectionNotification) error
+	}
+
+	// NotificationQueue implements a smart queue
+	NotificationQueue struct {
+		retryDelay  time.Duration
+		metrics     Metrics
+		done        chan struct{}
+		queue       chan *ProjectionNotification
+		queueBuffer int
+	}
+)
+
+func newNotificationQueue(queueBuffer int, retryDelay time.Duration, metrics Metrics) *NotificationQueue {
+	if retryDelay == 0 {
+		retryDelay = time.Millisecond * 50
+	}
+
+	return &NotificationQueue{
+		retryDelay:  retryDelay,
+		metrics:     metrics,
+		queueBuffer: queueBuffer,
+	}
+}
+
+// Channel returns the queue channel
+func (nq *NotificationQueue) Channel() chan *ProjectionNotification {
+	return nq.queue
+}
+
+// Close closes the queue channel
+func (nq *NotificationQueue) Close() {
+	close(nq.queue)
+}
+
+// Empty returns whether the queue is empty
+func (nq *NotificationQueue) Empty() bool {
+	return len(nq.queue) == 0
+}
+
+// PutBack sends a notification to the queue channel without further ado
+func (nq *NotificationQueue) PutBack(notification *ProjectionNotification) {
```

Comment:
If we could move the [logic regarding the timeout](https://github.com/hellofresh/goengine/pull/41/files#diff-6a6ba4cfc9a5e4a9e7d0aac85e41f166R131) into `Channel()` we could remove this func or make it private